crates/sui/src/client_ptb/builder.rs

```diff
@@ -0,0 +1,993 @@
+                let key = format!(
+                    "{}.{}",
+                    head.value,
+                    fields
+                        .iter()
+                        .map(|f| f.value.clone())
+                        .collect::<Vec<_>>()
+                        .join(".")
+                );
+                if let Some(addr) = self.addresses.get(&key) {
+                    self.addresses.insert(ident, *addr);
+                }
+            }
+            _ => (),
+        }
+    }
+
+    async fn get_protocol_config(&self, loc: Span) -> PTBResult<ProtocolConfig> {
+        let config = self
+            .reader
+            .get_protocol_config(None)
+            .await
+            .map_err(|e| err!(loc, "{e}"))?;
+        let chain_id = self
+            .reader
+            .get_chain_identifier()
+            .await
+            .map_err(|e| err!(loc, "{e}"))?;
+        Ok(if chain_id == get_mainnet_chain_identifier().to_string() {
+            ProtocolConfig::get_for_version(config.protocol_version, Chain::Mainnet)
+        } else if chain_id == get_testnet_chain_identifier().to_string() {
+            ProtocolConfig::get_for_version(config.protocol_version, Chain::Testnet)
+        } else {
+            ProtocolConfig::get_for_version(config.protocol_version, Chain::Unknown)
+        })
+    }
+
+    /// Resolve an object ID to a Move package.
+    async fn resolve_to_package(
+        &mut self,
+        package_id: ObjectID,
+        loc: Span,
+    ) -> PTBResult<MovePackage> {
+        let object = self
+            .reader
+            .get_object_with_options(package_id, SuiObjectDataOptions::bcs_lossless())
+            .await
+            .map_err(|e| err!(loc, "{e}"))?
+            .into_object()
+            .map_err(|e| err!(loc, "{e}"))?;
+        let Some(SuiRawData::Package(package)) = object.bcs else {
+            error!(
+                loc,
+                "BCS field in object '{}' is missing or not a package.", package_id
+            );
+        };
+        let protocol_config = self.get_protocol_config(loc).await?;
+        let package: MovePackage = MovePackage::new(
+            package.id,
+            object.version,
+            package.module_map,
+            protocol_config.max_move_package_size(),
+            package.type_origin_table,
+            package.linkage_table,
+        )
+        .map_err(|e| err!(loc, "{e}"))?;
+        Ok(package)
+    }
+
+    /// Resolves the argument to the move call based on the type information of the function being
+    /// called.
+    async fn resolve_move_call_arg(
+        &mut self,
+        view: &BinaryIndexedView<'_>,
+        ty_args: &[TypeTag],
+        sp!(loc, arg): Spanned<PTBArg>,
+        param: &SignatureToken,
+    ) -> PTBResult<Tx::Argument> {
+        // See if we've already resolved this argument or if it's an unambiguously pure value
+        if let Ok(res) = self.resolve(loc.wrap(arg.clone()), NoResolution).await {
```

Comment:
What does the `NoResolution` here mean? Should I read this like - try to resolve and if not, assign it to a `NoResolution` type?

Reply:
So this is gone now...but basically this was a way for us to query the builder of "have you already resolved this argument to a PTB input?" If it did no method (other than `pure` in the resolution context would be called). But, if it had not been resolved, then `resolve_object_id` would be called, and all the method in `NoResolution` does is error. So if we get an OK, then it would have either been a pure or already resolved to an object, if we get an `Err` then we know it was a (possible) object argument that had not been previously resolved.